### DISCUSSION TITLE
Is it possible to configure the code tag used for `mermaid` diagrams?

### DISCUSSION THREAD

**Discussion by matthias-colt** [gollum/gollum#1883] [General]
I'm trying to render wiki pages created in Azure DevOps using `gollum` (for off-line usage), which works quite good in principle.
Only small issue is that the wiki in Azure DevOps uses a different tag for `mermaid` diagrams.

Azure DevOps uses three colons `:::` instead of three backticks ` ``` `

Is there any option to configure the diagram tag for mermaid or any other workaround?

Azure DevOps `mermaid` example:

```text
::: mermaid
sequenceDiagram
    Christie->>Josh: Hello Josh, how are you?
    Josh-->>Christie: Great!
    Christie->>Josh: See you later!
:::
```

Gollum `mermaid` example:

````text
```mermaid
sequenceDiagram
    Christie->>Josh: Hello Josh, how are you?
    Josh-->>Christie: Great!
    Christie->>Josh: See you later!
```
````

**Comment by dometto** [comment]
The triple-backtick behavior is defined here: https://github.com/gollum/gollum-lib/blob/master/lib/gollum-lib/filter/code.rb#L50

As you can see the regex used to detect backtick-codeblocks is not abstracted very neatly (or at all). We could improve on this by:

1. defining a class variable in `Gollum::Filter::Code`, e.g. `@@codeblock_definitions`, that is an array of Regexes
2. Running the `data.gsub!` line for every regex in that array

Then you could override/add to `@@codeblock_definitions` in your `config.rb`. Would that seem like a good enough solution?

If the `:::` syntax is not just specific to Azure DevOp but more universally adopted, we could also consider letting Gollum support it out of the box.

**Reply by matthias-colt** [reply]
@dometto thanks for your quick reply and pointing me to the related code section.

Currently I have no idea where else the triple colon is used for this purpose. besides MS related like VS Code markdown renderer.

One issue might be that it should not change the code block tags in general but only for rendering mermaid diagrams. For "real" code the Azure DevOps wiki also uses the three backticks. If I correctly  understand the code you references is for identifying code blocks in general and not specific to a language like mermaid.

**Reply by dometto** [reply]
> For "real" code the Azure DevOps wiki also uses the three backticks. If I correctly understand the code you references is for identifying code blocks in general and not specific to a language like mermaid.

Ah I see, that at least speaks against adding support for `:::` to Gollum out of the box. Still, the solution could work fine as a customization option for individual users who just want to use `:::mermaid` (you could just refrain from using `:::` for anything else).

Alternatively, we would have to reimplement support for Mermaid as a filter (instead of the current solution based on [language handlers](https://github.com/gollum/gollum/wiki/Custom-language-handlers)). The reason for not doing that currently is performance: we would be walking through the document in search of codeblocks twice (once in the Mermaid filter and once in the Code filter) instead of once.

**Reply by matthias-colt** [reply]
I fully understand that the change currently doesn't make sense for Gollum.
For my use case I can do some workaround by replacing the `:::` with ` ``` ` in my offline copy with a script.
